src/main/java/com/conveyal/taui/analysis/broker/EC2RequestConfiguration.java

```diff
@@ -88,40 +88,48 @@ public String toString() {
         }
     }
 
-    RequestSpotInstancesRequest prepareSpotRequest(){
-        LaunchSpecification launchSpecification = new LaunchSpecification()
-                .withImageId(AnalysisServerConfig.workerAmiId)
-                .withInstanceType(AnalysisServerConfig.workerInstanceType)
-                .withSubnetId(AnalysisServerConfig.workerSubnetId)
-                .withIamInstanceProfile(new IamInstanceProfileSpecification().withArn(AnalysisServerConfig.workerIamRole))
-                .withUserData(userDataScript)
-                .withEbsOptimized(true);
+    RequestSpotInstancesRequest prepareSpotRequest (int nSpot, String clientToken) {
+        RequestSpotLaunchSpecification launchSpecification = RequestSpotLaunchSpecification.builder()
+                .imageId(AnalysisServerConfig.workerAmiId)
+                .instanceType(AnalysisServerConfig.workerInstanceType)
+                .subnetId(AnalysisServerConfig.workerSubnetId)
+                .iamInstanceProfile(IamInstanceProfileSpecification.builder().arn(AnalysisServerConfig.workerIamRole).build())
+                .userData(userDataScript)
+                .ebsOptimized(true)
+                .build();
 
-        return new RequestSpotInstancesRequest()
-                .withLaunchSpecification(launchSpecification);
+        return RequestSpotInstancesRequest.builder()
+                .launchSpecification(launchSpecification)
+                .instanceCount(nSpot)
+                .clientToken(clientToken)
+                .build();
     }
 
-    RunInstancesRequest prepareOnDemandRequest(){
+    RunInstancesRequest prepareOnDemandRequest (int minInstances, int maxInstances, String clientToken) {
         // Set tags so we can identify the instances immediately in the EC2 console. The worker user data makes
         // workers tag themselves, but with a lag.
-        TagSpecification instanceTags = new TagSpecification().withResourceType(ResourceType.Instance)
-                .withTags(
-                        new Tag("Name", "Analysis Worker"),
-                        new Tag("Project", "Analysis"),
-                        new Tag("networkId", category.graphId),
-                        new Tag("workerVersion", category.workerVersion),
-                        new Tag("group", group),
-                        new Tag("user", user)
-                );
+        TagSpecification instanceTags = TagSpecification.builder().resourceType(ResourceType.INSTANCE).tags(
+            Tag.builder().key("Name").value("Analysis Worker").build(),
+            Tag.builder().key("Project").value("Analysis").build(),
+            Tag.builder().key("networkId").value(category.graphId).build(),
+            Tag.builder().key("workerVersion").value(category.workerVersion).build(),
+            Tag.builder().key("group").value(group).build(),
+            Tag.builder().key("user").value(user).build()
```

Comment:
While we're making changes to tagging, it might also be worth adding region/project name/id tags, so we can more easily communicate to users a breakdown of their computation by region/project.

Reply:
I agree, that would be useful. We might also want to revise the way the tagging is performed though, since it's done in two different ways for spot and on-demand instances. It feels like a long term maintenance problem to be changing the startup script on the worker, keeping it in sync with all the tags that are set on the TagSpecification for the on demand request. If we had a more sophisticated Cluster Manager on the backend, it could perform the tagging once it saw that new instances had come online. As long as we're using the current system, maybe we could pass the TagSpecification object into the prepareSpotRequest, and it could somehow substitute that arbitrary set of tags into the startup script without caring/knowing what the individual tags were.

Reply:
Done! In the way described in my previous comment, making the tag command from the TagSpecification.